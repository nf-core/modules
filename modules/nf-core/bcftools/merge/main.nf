process BCFTOOLS_MERGE {
    tag "$meta.id"
    label 'process_medium'

    conda (params.enable_conda ? "bioconda::bcftools=1.16" : null)
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bcftools:1.16--hfe4b78e_1':
        'quay.io/biocontainers/bcftools:1.16--hfe4b78e_1' }"

    input:
    tuple val(meta), path(vcfs), path(tbis)
    path bed
    path fasta
    path fasta_fai

    output:
    tuple val(meta), path("*.{bcf,vcf}{,.gz}"), emit: merged_variants
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args   ?: ''
    def prefix   = task.ext.prefix ?: "${meta.id}"

    def regions = bed ? "--regions-file $bed" : ""
    def extension = args.contains("--output-type b") || args.contains("-Ob") ? "bcf.gz" :
                    args.contains("--output-type u") || args.contains("-Ou") ? "bcf" :
                    args.contains("--output-type v") || args.contains("-Ov") ? "vcf" :
                    "vcf.gz"

    """
    bcftools merge \\
        $regions \\
        --threads $task.cpus \\
        --output ${prefix}.${extension} \\
        $args \\
        *.vcf.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args   ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extension = args.contains("--output-type b") || args.contains("-Ob") ? "bcf.gz" :
                    args.contains("--output-type u") || args.contains("-Ou") ? "bcf" :
                    args.contains("--output-type v") || args.contains("-Ov") ? "vcf" :
                    "vcf.gz"
    """
    touch ${prefix}.${extension}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """
}
