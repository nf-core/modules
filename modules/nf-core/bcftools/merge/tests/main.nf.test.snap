{
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - tbi index": {
        "content": [
            "e0de448dc8e712956a03ce68d79a0b3a",
            "test.vcf.gz.tbi",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T12:34:16.977726522"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf output - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:17:34.532910365"
    },
    "sarscov2 - [vcf, tbi], [], [], bed": {
        "content": [
            "febdcfb851dcfc83d8248520830aef10",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:29:48.630057872"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - index - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:17:59.983157569"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf output": {
        "content": [
            "57bb84274f336465d0a0946b532093b0",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:29:05.528412678"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - bcf.gz output - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T11:58:46.619657457"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - tbi index - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:18:12.848227353"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T11:23:15.794389239"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - csi index": {
        "content": [
            "e0de448dc8e712956a03ce68d79a0b3a",
            "test.vcf.gz.csi",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T11:57:16.850641473"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output": {
        "content": [
            "e0de448dc8e712956a03ce68d79a0b3a",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T11:56:27.949031071"
    },
    "sarscov2 - [vcf, tbi], [], [], bed - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:18:19.273064822"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - bcf output": {
        "content": [
            "test.bcf",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T11:56:38.567500859"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - bcf output - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:17:47.356328326"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - one sample - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T15:13:18.708495878"
    },
    "homo_sapiens - [vcf, tbi], fasta, fai, bed  - vcf.gz output - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:56:45.706125286"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - index": {
        "content": [
            "e0de448dc8e712956a03ce68d79a0b3a",
            "test.vcf.gz.csi",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T12:33:43.639646108"
    },
    "homo_sapiens - [vcf, tbi], fasta, fai, bed  - vcf.gz output": {
        "content": [
            "645b7f7f9131bfe350a9ec3cf82c17fe",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:55:44.299812124"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - one sample": {
        "content": [
            "2a374cf02f0c32cf607646167e7f153b",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T15:37:38.491844702"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:17:28.188178904"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - vcf.gz output - csi index - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T13:18:06.430943593"
    },
    "sarscov2 - [vcf, tbi], [], [], []": {
        "content": [
            "e0de448dc8e712956a03ce68d79a0b3a",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T12:31:03.893007442"
    },
    "sarscov2 - [vcf, tbi], [], [], [] - bcf.gz output": {
        "content": [
            "test.bcf.gz",
            [
                "versions.yml:md5,d2c0a30d9a4cc6df89a464ae82e0c38a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-09-05T11:56:56.416361069"
    }
}